flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FactoryUtil.java

```diff
@@ -0,0 +1,570 @@
+	 *
+	 * <p>It considers {@link Catalog#getFactory()} if provided.
+	 */
+	public static DynamicTableSource createTableSource(
+			@Nullable Catalog catalog,
+			ObjectIdentifier objectIdentifier,
+			CatalogTable catalogTable,
+			ReadableConfig configuration,
+			ClassLoader classLoader) {
+		final DefaultDynamicTableContext context = new DefaultDynamicTableContext(
+			objectIdentifier,
+			catalogTable,
+			configuration,
+			classLoader);
+		try {
+			final DynamicTableSourceFactory factory = getDynamicTableFactory(
+				DynamicTableSourceFactory.class,
+				catalog,
+				context);
+			return factory.createDynamicTableSource(context);
+		} catch (Throwable t) {
+			throw new ValidationException(
+				String.format(
+					"Unable to create a source for reading table '%s'.\n\n" +
+					"Table options are:\n\n" +
+					"%s",
+					objectIdentifier.asSummaryString(),
+					catalogTable.getOptions()
+						.entrySet()
+						.stream()
+						.map(e -> stringifyOption(e.getKey(), e.getValue()))
+						.sorted()
+						.collect(Collectors.joining("\n"))),
+				t);
+		}
+	}
+
+	/**
+	 * Creates a {@link DynamicTableSink} from a {@link CatalogTable}.
+	 *
+	 * <p>It considers {@link Catalog#getFactory()} if provided.
+	 */
+	public static DynamicTableSink createTableSink(
+			@Nullable Catalog catalog,
+			ObjectIdentifier objectIdentifier,
+			CatalogTable catalogTable,
+			ReadableConfig configuration,
+			ClassLoader classLoader) {
+		final DefaultDynamicTableContext context = new DefaultDynamicTableContext(
+			objectIdentifier,
+			catalogTable,
+			configuration,
+			classLoader);
+		try {
+			final DynamicTableSinkFactory factory = getDynamicTableFactory(
+				DynamicTableSinkFactory.class,
+				catalog,
+				context);
+			return factory.createDynamicTableSink(context);
+		} catch (Throwable t) {
+			throw new ValidationException(
+				String.format(
+					"Unable to create a sink for writing table '%s'.\n\n" +
+					"Table options are:\n\n" +
+					"%s",
+					objectIdentifier.asSummaryString(),
+					catalogTable.getOptions()
+						.entrySet()
+						.stream()
+						.map(e -> stringifyOption(e.getKey(), e.getValue()))
+						.sorted()
+						.collect(Collectors.joining("\n"))),
+				t);
+		}
+	}
+
+	/**
+	 * Creates a utility that helps in discovering formats and validating all options for a {@link DynamicTableFactory}.
+	 *
+	 * <p>The following example sketches the usage:
+	 * <pre>{@code
+	 * // in createDynamicTableSource()
+	 * helper = FactoryUtil.createTableFactoryHelper(this, context);
+	 * keyFormat = helper.discoverScanFormat(classloader, MyFormatFactory.class, KEY_OPTION, "prefix");
+	 * valueFormat = helper.discoverScanFormat(classloader, MyFormatFactory.class, VALUE_OPTION, "prefix");
+	 * helper.validate();
+	 * ... // construct connector with discovered formats
+	 * }</pre>
+	 *
+	 * <p>Note: This utility checks for left-over options in the final step.
+	 */
+	public static TableFactoryHelper createTableFactoryHelper(
```

Comment:
Shall we add a helper for `FormatFactory` too? Currently, we don't have a validation utility used in `ScanFormatFactory#createScanFormat` to also check for unconsumed option keys.

Reply:
The format is logic is a bit different from the top-level factory. It cannot check unconsumed option keys, but only has a read-only view for which it can access its own options. Otherwise we need to change `ReadableConfig` for listing all options and maybe also `DelegatingConfiguration`. I think functionality-wise it doesn't make a difference. It is sufficient that only the top-level checks for unconsumed options considering all nested factories.

Reply:
For example, if we have a kafka table with the following options:  

```sql
'connector' = 'kafka',
'topic' = 'test_topic',
'format' = 'json',
'format.fail-on-missing-field' = 'true',
'format.invalid-option' = 'true'
```

Kafka connector factory don't have `format.*` options in `requiredOptions` and `optionalOptions`, so the unconsumed options should be validated by format factory (i.e. the `format.invalid-option`). Otherwise, users can write arbitraty format options. 

Besides, current `TableFactoryHelper#validate` doesn't take the format options into account. Which means the validation will fail if any format option exist.

Reply:
`TableFactoryHelper#validate` does take format options into account. In the current implementation, it doesn't matter where a typo is:

```
fromat.fail-on-missing-field
format.fial-on-missing-field
```

In both cases the exception is the same, that there are unconsumed options. Not every nested factory must check that and the upper level as well, we can just do it once.

Reply:
Sorry, I misunderstood the logic of `discoverOptionalFormatFactory`. Thanks for the explanation. I don't have other concerns now.